source/applications/advanced/color_balance.py

```diff
@@ -0,0 +1,257 @@
+"""
+Balance color for RGB image.
+"""
+
+import datetime
+from pathlib import Path
+import numpy as np
+import matplotlib.pyplot as plt
+import zivid
+
+
+def _display_rgb(rgb, title):
+    """Display RGB image.
+
+    Args:
+        rgb: RGB image
+        title: Image title
+
+    Returns None
+
+    """
+    plt.figure()
+    plt.imshow(rgb)
+    plt.title(title)
+    plt.show(block=False)
+
+
+def _compute_mean_rgb(rgb, pixels):
+    """Compute mean RGB values.
+
+    Args:
+        rgb: RGB image
+        pixels: Number of central pixels for computation
+
+    Returns:
+        mean_red: Red channel mean value
+        mean_green: Green channel mean value
+        mean_blue: Blue channel mean value
+
+    """
+    height = np.shape(rgb)[0]
+    width = np.shape(rgb)[1]
+    red = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        0,
+    ]
+    green = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        1,
+    ]
+    blue = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        2,
+    ]
+
+    mean_red = np.mean(np.reshape(red, -1, []))
+    mean_green = np.mean(np.reshape(green, -1, []))
+    mean_blue = np.mean(np.reshape(blue, -1, []))
+
+    return mean_red, mean_green, mean_blue
+
+
+def _rgb_from_3d_capture(camera):
+    """Get RGB image from 3D capture.
+
+    Args:
+        camera: Zivid camera
+
+    Returns:
+        rgb: RGB image
+
+    """
+    frame = camera.capture()
+    point_cloud = frame.get_point_cloud().to_array()
+    rgb = np.dstack([point_cloud["r"], point_cloud["g"], point_cloud["b"]])
+
+    return rgb
+
+
+def _color_balance_calibration(camera):
+    """Balance color for RGB image by taking images of white surface (piece of paper, wall, etc.) in a loop.
+
+    Args:
+        camera: Zivid camera
+
+    Returns:
+        corrected_red_balance: Corrected red balance
+        corrected_blue_balance: Corrected blue balance
+
+    """
+    print(f"Starting color balance calibration")
+    default_red_balance = camera.settings.red_balance
+    default_blue_balance = camera.settings.blue_balance
+    with camera.update_settings() as updater:
+        updater.settings.blue_balance = 1.0
+        updater.settings.red_balance = 1.0
+
+    rgb = _rgb_from_3d_capture(camera)
+    [mean_red, mean_green, mean_blue] = _compute_mean_rgb(rgb, 100)
+    print(
+        f"Mean color values: R =",
+        int(mean_red),
+        f"G =",
+        int(mean_green),
+        f"B =",
+        int(mean_blue),
+    )
+    _display_rgb(rgb, "RGB image before color balance (3D)")
+
+    while int(mean_green) != int(mean_red) and int(mean_green) != int(mean_blue):
```

Comment:
Is it neccesary to retake images? I think the function should just have a check that the exposure time is about right. 

You will end up with bad calibration regardless now if the user f.ex has an overexposed image.

Reply:
If the image is overexposed that will indeed lead to a bad calibration.

I don't understand the first comment. The reason it is necessary to retake images is that a single iteration is not enough to get rgain and bgain that make R G B channel mean values equal. Or are you suggesting instead to apply gains on the R G B channels, recalculate their mean, compare them, repeat?

Are you suggesting to add a capture loop that finds the right exposure?

Reply:
But why is one iteration not enought?

Reply:
It is nearly enough. the difference in R G B after one iteration is around somewhere between 1 and 3 (0 to 255). I don't understand why this happens though.